rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+two files (instead of 24 files) for each day of data, e.g., `base_file_13.parquet` and `base_file_14.parquet` containing
+the data for 2022-10-01 12-2 PM.
+
+Currently, such a fine-grained data skipping based on a function on a column cannot be achieved in Hudi, because
+transforming the `timestamp` to the hour of day is not order-preserving, thus the file pruning cannot directly leverage
+the file-level column stats of the original column of `timestamp`. In this case, Hudi has to scan all the files for a
+day and push the predicate down when reading parquet files, increasing the amount of data to be scanned.
+
+### Use Case 2: Support for different storage layouts and view partition as index
+
+Today, partitions are mainly viewed as a query optimization technique and partition pruning certainly helps to improve 
+query performance. However, if we think about it, partitions are really a storage optimization technique. Partitions
+help you organize the data for your convenience, while balancing cloud storage scaling issues (e.g. throttling or having
+too many files/objects under one path). From a query optimization perspective, partitions are really just a coarse index. 
+We can achieve the same goals as partition pruning with indexes.
+
+In this RFC, we propose how data is partitioned (hive-style, hashed/random prefix for cloud throttling) can be decoupled
+from how the data is queried. There can be different layouts:
+
+1. Files are stored under a base path, partitioned hive-style.
+2. Files are stored under random prefixes attached to a base path, still hive-style partitioned (RFC-60) e.g. `s3://<
+   random-prefix1>path/to/table/partition1=abc`, `s3://<random_prefix2>path/to/table/partition1=xyz`.
+3. Files are stored across different buckets completely scattered on cloud storage e.g. `s3://a/b/c/f1`, `s3://x/y/f2`.
+4. Partitions can evolve. For instance, you have an old Hive table which is horribly partitioned, can we ensure that the
+   new data is not only partitioned well but queries able to efficiently skip data without rewriting the old data with
+   the new partition spec.
+
+Consider a case where event logs are stream from microservices and ingested into a raw event table. Each event log
+contains a `timestamp` and an associated organization ID (`org_id`). Most queries on the table are organization specific
+and fetch logs for a particular time range. A user may attempt to physically partition the data by both `org_id`
+and `date(timestamp)`. If there are 1K organization IDs and one year of data, such a physical partitioning scheme writes
+at least `365 days x 1K IDs = 365K` data files under 365K partitions. In most cases, the data can be highly skewed based
+on the organizations, with most organizations having less data and a handful of organizations having the majority of the
+data, so that there can be many small data files. In such a case, the user may want to evolve the partitioning by
```

Comment:
this is still a big anti-pattern where the table will have uneven performance. We should still advice the users to fix the partitioning across the board, given there are effective/fast techniques at our disposal now.

Reply:
+1 on considering this scenario as well, for completeness.